src/coreclr/src/jit/compiler.h

```diff
@@ -8815,6 +8828,8 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
         uint64_t compSupportsISA;
```

Comment:
Can we add a brief comment explaining the difference between these 3?

Is it currently (IIRC, please correct anything needed):
* `compSupportsISA`: The compiler is configured to emit such instructions, the target hardware may or may not support it
* `compSupportsISAReported`: Indicates some method is utilizing the given ISA in some code path
*
* `compSupportsISAExactly`: Indicates the VM has said an ISA is supported by the target hardware?

The names aren't super clear and at least for me it gets hard to mentally track which bits do what:
* `compSupportsHWIntrinsic` reports `COMPlus_EnableHWIntrinsic != 0`, `COMPlus_FeatureSIMD != 0` (basically, the compiler reports `true` provided the ISA isn't partially implemented and `HWIntrinsics` aren't generally disabled)
* `compExactlyDependsOn` used to report usage and return `compSupportsISA`; not looks to report usage and return `compSupportsISAExactly` which is dependent on `notifyInstructionSetUsage` returning `true`
* `compOpportunisticallyDependsOn` returns `compExactlyDependsOn` but only if `compSupportsISA` is true, effectively allowing code paths to not report usage if they can codify an alternative instead
*
* `compHWIntrinsicDependsOn` effectively returns `compSupportsISA` but additionally calls `compExactlyDependsOn` if true

Reply:
It's not clear to me exactly why we need `compSupportsISAExactly` for AOT scenarios rather than just having the VM accurately report `compSupportsISA`?
Likewise, for JIT scenarios (at least for S.P.Corelib), I would expect all of our usages are either behind a user provided `IsSupported` call or behind a `compOpportunisticallyDependsOn` check

I had the understanding that in the case of the latter, the method would be marked as using `SSE4.1`, for example, and if at runtime the method did not actually support it, it would be rejitted and get told the actual ISAs supported

Reply:
> It's not clear to me exactly why we need compSupportsISAExactly for AOT scenarios rather than just having the VM accurately report compSupportsISA?

For NativeAOT that does not have fallback to JIT, we want to produce code that both works on all platforms and also leverages hardware intrinsics as much as possible. We achieve that by substituting `ISA.IsSupported` with a dynamic check like `bool Popcnt.IsSupport => (g_flags & PopcntBit) != 0;` (this dynamic check is provided by the VM as IL).

This scheme only works for explicit hardware intrinsics that are protected by explicit checks. It does not work for cases where the JIT chooses to use hardware intrinsics implicitly.

Reply:
I have updated the comments. Please take a look.